File: app/routers/connections.py
"""Connection API routes - /api/my/connections/* endpoints."""

import asyncio
import logging
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse

from app.utils.helpers import _sanitize_error, generate_vpn_link, get_ssh, get_protocol_manager
from config import get_db
from dependencies import get_current_user
from schemas import MyAddConnectionRequest

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/my/connections")


@router.get("")
async def api_my_connections(request: Request, user: dict = Depends(get_current_user)):
    db = get_db()
    conns = db.get_connections_by_user(user["id"])
    for c in conns:
        sid = c.get("server_id", 0)
        srv = db.get_server_by_id(sid)
        if srv:
            c["server_name"] = srv.get("name", srv.get("host", ""))
        else:
            c["server_name"] = "Unknown"

    # Include effective limits for the frontend
    settings = db.get_all_settings()
    global_limits = settings.get("limits", {})
    user_limits = user.get("limits", {})
    effective_limits = {
        "max_connections": user_limits.get(
            "max_connections_per_user", global_limits.get("max_connections_per_user", 10)
        ),
        "current_connections": len(conns),
    }

    return {"connections": conns, "limits": effective_limits}


@router.post("/add")
async def api_my_add_connection(
    request: Request, req: MyAddConnectionRequest, user: dict = Depends(get_current_user)
):

    # Validate user account status
    if not user.get("enabled", True):
        return JSONResponse({"error": "Account is disabled"}, status_code=403)

    # Check expiration
    exp_str = user.get("expiration_date")
    if exp_str:
        try:
            exp_date = datetime.fromisoformat(exp_str)
            if datetime.now() > exp_date:
                return JSONResponse({"error": "Account expired"}, status_code=403)
        except Exception:
            pass  # Invalid date format, ignore

    # Check traffic limit
    traffic_limit = user.get("traffic_limit", 0)
    traffic_used = user.get("traffic_used", 0)
    if traffic_limit > 0 and traffic_used >= traffic_limit:
        return JSONResponse({"error": "Traffic limit exceeded"}, status_code=403)

    # ---- Rate Limiting & Connection Limits ----
    db = get_db()

    # Resolve limits
    settings = db.get_all_settings()
    global_limits = settings.get("limits", {})
    user_limits = user.get("limits", {})
    max_conns_per_user = user_limits.get(
        "max_connections_per_user", global_limits.get("max_connections_per_user", 10)
    )
    rate_limit_count = user_limits.get(
        "connection_rate_limit_count", global_limits.get("connection_rate_limit_count", 5)
    )
    rate_limit_window = user_limits.get(
        "connection_rate_limit_window", global_limits.get("connection_rate_limit_window", 60)
    )

    # Check per-user connection count
    user_conns = db.get_connections_by_user(user["id"])
    if len(user_conns) >= max_conns_per_user:
        logger.warning(
            f"Rate limit triggered (max connections): user_id={user['id']}, "
            f"current={len(user_conns)}, limit={max_conns_per_user}"
        )
        return JSONResponse(
            {
                "error": f"Maximum connections limit reached ({max_conns_per_user})",
                "limit": max_conns_per_user,
                "current": len(user_conns),
            },
            status_code=428,
        )

    # Check time-based rate limiting (sliding window)
    now = datetime.now()
    recent = db.get_recent_connections_log(user["id"], rate_limit_window)
    if len(recent) >= rate_limit_count:
        oldest = min(recent, key=lambda e: e["created_at"])
        try:
            oldest_ts = datetime.fromisoformat(oldest["created_at"])
            retry_after = int(rate_limit_window - (now - oldest_ts).total_seconds()) + 1
        except Exception:
            retry_after = rate_limit_window
        logger.warning(
            f"Rate limit triggered (sliding window): user_id={user['id']}, "
            f"recent_count={len(recent)}, limit={rate_limit_count}, window={rate_limit_window}s"
        )
        return JSONResponse(
            {
                "error": f"Connection rate limit exceeded ({rate_limit_count} per {rate_limit_window}s)",
                "retry_after": retry_after,
            },
            status_code=428,
            headers={"Retry-After": str(retry_after)},
        )

    # Validate server exists
    server = db.get_server_by_id(req.server_id)
    if server is None:
        return JSONResponse({"error": "Server not found"}, status_code=404)

    # Verify protocol is installed
    proto_info = server.get("protocols", {}).get(req.protocol, {})
    if not proto_info or not proto_info.get("installed", False):
        return JSONResponse(
            {"error": f"Protocol {req.protocol} is not installed on this server"},
            status_code=400,
        )

    # Check for duplicate connection name
    existing_names = {c.get("name", "") for c in user_conns}
    if req.name in existing_names:
        return JSONResponse(
            {
                "error": "duplicate_name",
                "message": "A connection with this name already exists.",
            },
            status_code=409,
        )

    port = proto_info.get("port", "55424")

    # Prune old connection log entries
    db.prune_connection_log(1000)

    ssh = None
    try:
        ssh = get_ssh(server)
        await asyncio.to_thread(ssh.connect)
        manager = get_protocol_manager(ssh, req.protocol)

        # Create client on remote server
        if req.protocol == "telemt":
            result = await asyncio.to_thread(
                manager.add_client,
                req.protocol,
                req.name,
                server["host"],
                port,
                telemt_quota=req.telemt_quota,
                telemt_max_ips=req.telemt_max_ips,
                telemt_expiry=req.telemt_expiry,
            )
        else:
            result = await asyncio.to_thread(
                manager.add_client, req.protocol, req.name, server["host"], port
            )

        if result.get("client_id"):
            new_conn = {
                "id": str(uuid.uuid4()),
                "user_id": user["id"],
                "server_id": req.server_id,
                "protocol": req.protocol,
                "client_id": result["client_id"],
                "name": req.name,
                "created_at": now.isoformat(),
            }
            db.create_connection(new_conn)
            db.log_connection_creation(user["id"])

            # Enrich connection with server_name for frontend
            new_conn["server_name"] = server.get("name", server.get("host", "Unknown"))

            # Build response
            response = {
                "status": "success",
                "connection": new_conn,
                "client_id": result["client_id"],
            }
            if result.get("config"):
                response["config"] = result["config"]
                response["vpn_link"] = generate_vpn_link(result["config"])
            return response
        else:
            return JSONResponse({"error": "Failed to create connection on server"}, status_code=500)
    except Exception as e:
        logger.exception("Error in api_my_add_connection")
        safe_msg = _sanitize_error(str(e), "Failed to create connection")
        return JSONResponse({"error": safe_msg}, status_code=500)
    finally:
        if ssh:
            await asyncio.to_thread(ssh.disconnect)


@router.post("/{connection_id}/config")
async def api_my_connection_config(
    request: Request, connection_id: str, user: dict = Depends(get_current_user)
):
    try:
        db = get_db()
        conn = db.get_connection_by_id(connection_id)
        if not conn or conn.get("user_id") != user["id"]:
            return JSONResponse({"error": "Connection not found"}, status_code=404)
        sid = conn["server_id"]
        server = db.get_server_by_id(sid)
        if server is None:
            return JSONResponse({"error": "Server not found"}, status_code=404)
        proto_info = server.get("protocols", {}).get(conn["protocol"], {})
        port = proto_info.get("port", "55424")
        ssh = get_ssh(server)
        await asyncio.to_thread(ssh.connect)
        # Use appropriate manager for the protocol
        manager = get_protocol_manager(ssh, conn["protocol"])
        config = await asyncio.to_thread(
            manager.get_client_config,
            conn["protocol"],
            conn["client_id"],
            server["host"],
            port,
        )
        await asyncio.to_thread(ssh.disconnect)
        vpn_link = generate_vpn_link(config) if config else ""
        return {"config": config, "vpn_link": vpn_link}
    except Exception as e:
        logger.exception("Error getting my connection config")
        return JSONResponse({"error": _sanitize_error(str(e))}, status_code=500)

File: amnezia-web-ui-go/internal/handlers/connections.go
package handlers

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/devops-igor/amnezia-web-ui-go/internal/models"
	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
)

// effectiveMaxConnectionsPerUser returns the per-user override of the connection limit
// when set, otherwise the global limit.
func effectiveMaxConnectionsPerUser(user *models.User, global int) int {
	if user != nil && user.Limits != nil {
		if v, ok := user.Limits["max_connections_per_user"].(float64); ok && int(v) > 0 {
			return int(v)
		}
	}
	return global
}

// effectiveRateLimit returns the per-user override of the connection rate limit
// count/window when set, otherwise the global values.
func effectiveRateLimit(user *models.User, count, window int) (int, int) {
	if user != nil && user.Limits != nil {
		if v, ok := user.Limits["connection_rate_limit_count"].(float64); ok && int(v) > 0 {
			count = int(v)
		}
		if v, ok := user.Limits["connection_rate_limit_window"].(float64); ok && int(v) > 0 {
			window = int(v)
		}
	}
	return count, window
}

// serverReachabilityInfo resolves cached reachability for a server: status string and boolean flag.
func (h *Handlers) serverReachabilityInfo(ctx context.Context, serverID int64) (string, bool) {
	status, err := h.db.GetServerStatus(ctx, serverID)
	if err != nil {
		return "unknown", false
	}
	switch status {
	case models.ReachabilityOnline:
		return string(status), true
	case models.ReachabilityOffline:
		return string(status), false
	default:
		return string(models.ReachabilityUnknown), false
	}
}

// UserGetMyConnectionsHandler returns the authenticated user's active connections.
func (h *Handlers) UserGetMyConnectionsHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	ctx := r.Context()
	conns, err := h.db.GetConnectionsByUserID(ctx, sess.UserID)
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to retrieve connections")
		return
	}

	servers, _ := h.db.GetAllServers(ctx)
	serversMap := make(map[int64]string)
	for _, s := range servers {
		serversMap[s.ID] = s.Name
	}

	enriched := make([]map[string]any, len(conns))
	for i, c := range conns {
		srvName := serversMap[c.ServerID]
		if srvName == "" {
			srvName = fmt.Sprintf("Server #%d", c.ServerID)
		}

		srvStatus, srvReachable := h.serverReachabilityInfo(ctx, c.ServerID)

		enriched[i] = map[string]any{
			"id":               c.ID,
			"user_id":          c.UserID,
			"server_id":        c.ServerID,
			"server_name":      srvName,
			"protocol":         c.Protocol,
			"client_id":        c.ClientID,
			"name":             c.Name,
			"awg_mimicry":      c.AWGMimicry,
			"traffic_total_rx": c.TrafficTotalRx,
			"traffic_total_tx": c.TrafficTotalTx,
			"traffic_total":    c.TrafficTotal,
			"created_at":       c.CreatedAt.Format(time.RFC3339),
			"server_reachable": srvReachable,
			"server_status":    srvStatus,
		}
	}

	var limits models.ConnectionLimits
	_ = h.db.GetSetting(ctx, "limits", &limits)
	if limits.MaxConnectionsPerUser <= 0 {
		limits.MaxConnectionsPerUser = 10
	}

	user, _ := h.db.GetUser(ctx, sess.UserID)
	effectiveMax := effectiveMaxConnectionsPerUser(user, limits.MaxConnectionsPerUser)

	h.JSON(w, http.StatusOK, map[string]any{
		"status":      "ok",
		"connections": enriched,
		"limits": map[string]any{
			"max_connections":     effectiveMax,
			"current_connections": len(conns),
		},
	})
}

// UserAddConnectionHandler provisions a new client connection for the authenticated user.
func (h *Handlers) UserAddConnectionHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	unlock := h.lockUser(sess.UserID)
	defer unlock()

	var req models.MyAddConnectionRequest
	if err := h.DecodeJSON(r, &req); err != nil {
		h.JSONError(w, http.StatusBadRequest, "validation_failed", "Invalid request body")
		return
	}

	if err := req.Validate(); err != nil {
		h.JSONError(w, http.StatusBadRequest, "validation_failed", err.Error())
		return
	}

	ctx := r.Context()
	user, err := h.db.GetUser(ctx, sess.UserID)
	if err != nil || user == nil {
		h.JSONError(w, http.StatusNotFound, "not_found", "User not found")
		return
	}

	// 1. Account status checks
	if !h.checkUserEligible(w, user) {
		return
	}

	// 2. Limits and Rate Limiting (per-user overrides win over global settings)
	userConns, _ := h.db.GetConnectionsByUserID(ctx, user.ID)
	if !h.checkConnectionLimits(w, ctx, user, userConns) {
		return
	}

	// 3. Server and duplicate name check
	server, err := h.db.GetServer(ctx, req.ServerID)
	if err != nil || server == nil {
		h.JSONError(w, http.StatusNotFound, "not_found", "Server not found")
		return
	}

	if h.hasDuplicateConnectionName(userConns, req.Name) {
		h.JSON(w, http.StatusConflict, map[string]any{
			"error":   "duplicate_name",
			"message": "A connection with this name already exists.",
		})
		return
	}

	protoMgr, err := h.GetProtocolManager(req.Protocol)
	if err != nil {
		h.JSONError(w, http.StatusBadRequest, "invalid_protocol", err.Error())
		return
	}

	// Guard: the requested protocol must be installed on this server.
	if !isProtocolInstalled(server, req.Protocol) {
		h.JSONError(w, http.StatusBadRequest, "protocol_not_installed",
			fmt.Sprintf("Protocol %s is not installed on this server", req.Protocol))
		return
	}

	clientParams := map[string]any{
		"name": req.Name,
	}
	h.appendConnectionParams(clientParams, req)

	result, err := protoMgr.AddClient(ctx, server, clientParams)
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "add_client_failed", "Failed to create client on server")
		return
	}

	clientID, _ := result["client_id"].(string)
	if clientID == "" {
		clientID, _ = result["clientId"].(string)
	}

	configStr, _ := result["config"].(string)
	if configStr == "" && clientID != "" {
		configStr, _ = protoMgr.GetClientConfig(ctx, server, clientID)
	}
	vpnLink := GenerateVPNLink(configStr)

	newConn := &models.UserConnection{
		ID:         uuid.NewString(),
		UserID:     user.ID,
		ServerID:   req.ServerID,
		Protocol:   req.Protocol,
		ClientID:   clientID,
		Name:       req.Name,
		AWGMimicry: models.AWGMimicryAuto,
		CreatedAt:  time.Now(),
	}
	if req.AWGMimicry != nil {
		newConn.AWGMimicry = models.AWGMimicryProfile(*req.AWGMimicry)
	}

	if _, err := h.db.CreateConnection(ctx, newConn); err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to save connection record")
		return
	}

	_ = h.db.LogConnectionCreation(ctx, user.ID)

	h.audit(r, "connection.user_add", map[string]any{"user_id": user.ID, "server_id": req.ServerID, "protocol": req.Protocol, "client_id": clientID, "name": req.Name})
	h.JSON(w, http.StatusOK, map[string]any{
		"status":     "ok",
		"client_id":  clientID,
		"config":     configStr,
		"vpn_link":   vpnLink,
		"connection": newConn,
	})
}

// UserGetConnectionConfigHandler returns configuration text for an owned connection.
func (h *Handlers) UserGetConnectionConfigHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	connectionID := chi.URLParam(r, "connection_id")
	if connectionID == "" {
		h.JSONError(w, http.StatusBadRequest, "invalid_parameter", "connection_id is required")
		return
	}

	ctx := r.Context()
	conn, err := h.db.GetConnection(ctx, connectionID)
	if err != nil || conn == nil || conn.UserID != sess.UserID {
		h.JSONError(w, http.StatusNotFound, "not_found", "Connection not found")
		return
	}

	server, err := h.db.GetServer(ctx, conn.ServerID)
	if err != nil || server == nil {
		h.JSONError(w, http.StatusNotFound, "not_found", "Server not found")
		return
	}

	protoMgr, err := h.GetProtocolManager(conn.Protocol)
	if err != nil {
		h.JSONError(w, http.StatusBadRequest, "invalid_protocol", err.Error())
		return
	}

	configStr, err := protoMgr.GetClientConfig(ctx, server, conn.ClientID)
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to get config")
		return
	}

	vpnLink := GenerateVPNLink(configStr)
	filename := fmt.Sprintf("%s.conf", conn.Name)

	h.JSON(w, http.StatusOK, map[string]any{
		"status":      "ok",
		"config":      configStr,
		"filename":    filename,
		"vpn_link":    vpnLink,
		"awg_mimicry": conn.AWGMimicry,
	})
}

// UserGetConnectionKitHandler generates and serves a connection kit ZIP archive.
func (h *Handlers) UserGetConnectionKitHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	connectionID := chi.URLParam(r, "connection_id")
	if connectionID == "" {
		h.JSONError(w, http.StatusBadRequest, "invalid_parameter", "connection_id is required")
		return
	}

	ctx := r.Context()
	conn, err := h.db.GetConnection(ctx, connectionID)
	if err != nil || conn == nil || conn.UserID != sess.UserID {
		h.JSONError(w, http.StatusNotFound, "not_found", "Connection not found")
		return
	}

	server, err := h.db.GetServer(ctx, conn.ServerID)
	if err != nil || server == nil {
		h.JSONError(w, http.StatusNotFound, "not_found", "Server not found")
		return
	}

	protoMgr, err := h.GetProtocolManager(conn.Protocol)
	if err != nil {
		h.JSONError(w, http.StatusBadRequest, "invalid_protocol", err.Error())
		return
	}

	configStr, err := protoMgr.GetClientConfig(ctx, server, conn.ClientID)
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to get config")
		return
	}

	vpnLink := GenerateVPNLink(configStr)
	zipBytes, err := BuildConnectionKitZip(conn.Name, configStr, vpnLink)
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to build connection kit archive")
		return
	}

	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s-kit.zip\"", conn.Name))
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(zipBytes)
}

// UserRenameConnectionHandler renames an owned client connection.
func (h *Handlers) UserRenameConnectionHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	connectionID := chi.URLParam(r, "connection_id")
	if connectionID == "" {
		h.JSONError(w, http.StatusBadRequest, "invalid_parameter", "connection_id is required")
		return
	}

	var req models.RenameConnectionRequest
	if err := h.DecodeJSON(r, &req); err != nil {
		h.JSONError(w, http.StatusBadRequest, "validation_failed", "Invalid request body")
		return
	}

	if err := req.Validate(); err != nil {
		h.JSONError(w, http.StatusBadRequest, "validation_failed", err.Error())
		return
	}

	ctx := r.Context()
	conn, err := h.db.GetConnection(ctx, connectionID)
	if err != nil || conn == nil || conn.UserID != sess.UserID {
		h.JSONError(w, http.StatusNotFound, "not_found", "Connection not found")
		return
	}

	// Check duplicates
	userConns, err := h.db.GetConnectionsByUserID(ctx, sess.UserID)
	if err == nil {
		for _, c := range userConns {
			if c.ID != connectionID && strings.EqualFold(c.Name, req.Name) {
				h.JSON(w, http.StatusConflict, map[string]any{
					"error":   "duplicate_name",
					"message": "A connection with this name already exists.",
				})
				return
			}
		}
	}

	_, err = h.db.UpdateConnection(ctx, connectionID, map[string]any{
		"name": req.Name,
	})
	if err != nil {
		h.JSONError(w, http.StatusInternalServerError, "internal_error", "Failed to rename connection")
		return
	}

	h.audit(r, "connection.user_rename", map[string]any{"user_id": sess.UserID, "connection_id": connectionID, "old_name": conn.Name, "new_name": req.Name})
	h.JSON(w, http.StatusOK, map[string]any{
		"status": "ok",
		"name":   req.Name,
	})
}

// UserDeleteConnectionHandler removes an owned client connection.
func (h *Handlers) UserDeleteConnectionHandler(w http.ResponseWriter, r *http.Request) {
	sess := h.GetSession(r)
	if sess == nil || !sess.IsAuthenticated() {
		h.JSONError(w, http.StatusUnauthorized, "unauthorized", "Authentication required")
		return
	}

	connectionID := chi.URLParam(r, "connection_id")
	if connectionID == "" {
		h.JSONError(w, http.StatusBadRequest, "invalid_parameter", "connection_id is required")
		return
	}

	ctx := r.Context()
	conn, err := h.db.GetConnection(ctx, connectionID)
	if err != nil || conn == nil || conn.UserID != sess.UserID {
		h.JSONError(w, http.StatusNotFound, "not_found", "Connection not found")
		return
	}

	server, err := h.db.GetServer(ctx, conn.ServerID)
	if err == nil && server != nil {
		if protoMgr, err := h.GetProtocolManager(conn.Protocol); err == nil && protoMgr != nil {
			_ = protoMgr.RemoveClient(ctx, server, conn.ClientID)
		}
	}

	_, _ = h.db.DeleteConnection(ctx, connectionID)

	h.audit(r, "connection.user_delete", map[string]any{"user_id": sess.UserID, "connection_id": connectionID, "name": conn.Name})
	h.JSONOK(w)
}

// hasDuplicateConnectionName reports whether any existing connection already uses the given name (case-insensitive).
func (h *Handlers) hasDuplicateConnectionName(conns []models.UserConnection, name string) bool {
	for _, c := range conns {
		if strings.EqualFold(c.Name, name) {
			return true
		}
	}
	return false
}

// accountExpired reports whether the user account has passed either expiry timestamp.
func (h *Handlers) accountExpired(user *models.User) bool {
	now := time.Now()
	if user.ExpiresAt != nil && now.After(*user.ExpiresAt) {
		return true
	}
	return user.ExpirationDate != nil && now.After(*user.ExpirationDate)
}

// resolveConnectionLimits loads connection limits from settings, applying defaults for unset values.
func (h *Handlers) resolveConnectionLimits(ctx context.Context) models.ConnectionLimits {
	var limits models.ConnectionLimits
	_ = h.db.GetSetting(ctx, "limits", &limits)
	if limits.MaxConnectionsPerUser <= 0 {
		limits.MaxConnectionsPerUser = 10
	}
	if limits.ConnectionRateLimitCount <= 0 {
		limits.ConnectionRateLimitCount = 5
	}
	if limits.ConnectionRateLimitWindow <= 0 {
		limits.ConnectionRateLimitWindow = 60
	}
	return limits
}

// appendConnectionParams copies optional per-protocol client parameters from the request into
// the protocol-manager params map.
func (h *Handlers) appendConnectionParams(params map[string]any, req models.MyAddConnectionRequest) {
	if req.TelemtQuota != nil {
		params["telemt_quota"] = *req.TelemtQuota
	}
	if req.TelemtMaxIPs != nil {
		params["telemt_max_ips"] = *req.TelemtMaxIPs
	}
	if req.TelemtExpiry != nil {
		params["telemt_expiry"] = *req.TelemtExpiry
	}
	if req.AWGSpeedLimitDown != nil {
		params["speed_limit_down"] = *req.AWGSpeedLimitDown
	}
	if req.AWGSpeedLimitUp != nil {
		params["speed_limit_up"] = *req.AWGSpeedLimitUp
	}
	if req.AWGMimicry != nil {
		params["awg_mimicry"] = *req.AWGMimicry
	}
}

func (h *Handlers) checkUserEligible(w http.ResponseWriter, user *models.User) bool {
	if !user.Enabled {
		h.JSONError(w, http.StatusForbidden, "account_disabled", "Account is disabled")
		return false
	}
	if h.accountExpired(user) {
		h.JSONError(w, http.StatusForbidden, "account_expired", "Account expired")
		return false
	}
	if user.TrafficLimit > 0 && user.TrafficUsed >= user.TrafficLimit {
		h.JSONError(w, http.StatusForbidden, "traffic_limit_exceeded", "Traffic limit exceeded")
		return false
	}
	return true
}

func (h *Handlers) checkConnectionLimits(w http.ResponseWriter, ctx context.Context, user *models.User, userConns []models.UserConnection) bool {
	limits := h.resolveConnectionLimits(ctx)
	effectiveMax := effectiveMaxConnectionsPerUser(user, limits.MaxConnectionsPerUser)
	rateCount, rateWindow := effectiveRateLimit(user, limits.ConnectionRateLimitCount, limits.ConnectionRateLimitWindow)

	if len(userConns) >= effectiveMax {
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		h.JSON(w, http.StatusPreconditionRequired, map[string]any{
			"error":   "max_connections_reached",
			"detail":  fmt.Sprintf("Maximum connections limit reached (%d)", effectiveMax),
			"limit":   effectiveMax,
			"current": len(userConns),
		})
		return false
	}

	recentLogs, err := h.db.GetRecentConnectionsLog(ctx, user.ID, rateWindow)
	if err == nil && len(recentLogs) >= rateCount {
		retryAfter := rateWindow
		w.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfter))
		h.JSON(w, http.StatusPreconditionRequired, map[string]any{
			"error":       "connection_rate_limit_exceeded",
			"detail":      fmt.Sprintf("Connection rate limit exceeded (%d per %ds)", rateCount, rateWindow),
			"retry_after": retryAfter,
		})
		return false
	}
	return true
}
